timesketch/lib/utils.py

```diff
@@ -83,97 +90,99 @@ def _scrub_special_tags(dict_obj):
 
     Raises:
         RuntimeError: when there are missing fields.
         DataIngestionError: when there are issues with the data ingestion.
     """
-    # Columns that must be present in the CSV file.
-    mandatory_fields = ['message', 'datetime', 'timestamp_desc']
-
     # Ensures delimiter is a string.
     if not isinstance(delimiter, six.text_type):
         delimiter = codecs.decode(delimiter, 'utf8')
 
-    # Due to issues with python2.
-    if six.PY2:
-        delimiter = str(delimiter)
-
-    reader = csv.DictReader(file_handle, delimiter=delimiter)
-    csv_header = reader.fieldnames
-    missing_fields = []
-    # Validate the CSV header
-    for field in mandatory_fields:
-        if field not in csv_header:
-            missing_fields.append(field)
-    if missing_fields:
-        raise RuntimeError(
-            'Missing fields in CSV header: {0:s}'.format(
-                ','.join(missing_fields)))
+    header_reader = pandas.read_csv(file_handle, sep=delimiter, nrows=0)
+    _validate_csv_fields(mandatory_fields, header_reader)
     try:
-        for row in reader:
-            # There is a condition in which the CSV reader can read a
-            # single lines as multiple, causing issues with importing.
-            # TODO: Swap the CSV library for the user of pandas.
-            if not row:
-                continue
-            if not row['datetime']:
+        reader = pandas.read_csv(file_handle, sep=delimiter,
+                                 chunksize=DEFAULT_CHUNK_SIZE)
+        for chunk in reader:
+            skipped_rows = chunk[chunk['datetime'].isnull()]
+            for row in skipped_rows.iterrows():
                 logger.warning(
                     'Row missing a datetime object, skipping [{0:s}]'.format(
                         ','.join([str(x).replace(
                             '\n', '').strip() for x in row.values()])))
-                continue
-            try:
-                # normalize datetime to ISO 8601 format if it's not the case.
-                parsed_datetime = parser.parse(row['datetime'])
-                row['datetime'] = parsed_datetime.isoformat()
-
-                normalized_timestamp = int(
-                    time.mktime(parsed_datetime.utctimetuple()) * 1000000)
-                normalized_timestamp += parsed_datetime.microsecond
-                row['timestamp'] = str(normalized_timestamp)
-                if 'tag' in row:
-                    row['tag'] = [x for x in _parse_tag_field(row['tag']) if x]
 
-                _scrub_special_tags(row)
-            except ValueError:
-                continue
+            # Normalize datetime to ISO 8601 format if it's not the case.
+            chunk['datetime'] = pandas.to_datetime(chunk['datetime'])
 
-            yield row
-    except csv.Error as e:
+            chunk['timestamp'] = chunk['datetime'].apply(_get_unix_time)
+            chunk['datetime'] = chunk['datetime'].apply(pandas.Timestamp
```

Comment:
you can also just do 


```
chunk['datetime'] = chunk['datetime'].dt.strftime('%Y-%m-%dT%H:%M:%S%z')

Reply:
re: `chunk['datetime']`. Isn't it a bit clearer for future maintainers if we use the `Timestamp.isoformat` rather than the magic formatting string? Another concern is that `%z` seem to be [platform-dependent](https://stackoverflow.com/questions/32353015/python-time-strftime-z-is-always-zero-instead-of-timezone-offset).

Reply:
yes, I'm a fan of using isoformat, and yes I don't know why I added the `%z`, old habit, but I agree